### PR TITLE
Iox #33 implement tests for unix domain socket

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
This PR introduces unit tests for the unix domain sockets and abstracted the posix calls further. All socket related calls have an `iox_` prefix, this was required since in windows some calls cannot be mapped one to one.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- #33

### CODE REVIEW

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp [hunk 1]
```diff
@@ -204,7 +209,7 @@ cxx::expected<IpcChannelError> UnixDomainSocket::timedSend(const std::string& ms
                                                            const units::Duration& timeout) const noexcept
 {
     constexpr uint64_t NULL_TERMINATOR_SIZE = 1U;
-    if (msg.size() > m_maxMessageSize + NULL_TERMINATOR_SIZE)
+    if (msg.size() >= m_maxMessageSize + NULL_TERMINATOR_SIZE)
```
Comment (elBoberido) [inline iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:212]: Hmm, shouldn't this be just `msg.size > m_maxMessageSize`?

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp [hunk 2]
```diff
@@ -523,7 +522,7 @@ IpcChannelError UnixDomainSocket::convertErrnoToIpcChannelError(const int32_t er
 
 bool UnixDomainSocket::isNameValid(const UdsName_t& name) noexcept
 {
-    return !(name.empty() || name.size() < SHORTEST_VALID_NAME || name.size() > LONGEST_VALID_NAME);
```
Comment (elBoberido) [inline iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: Why is this removed?

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 3]
```diff
@@ -78,36 +125,229 @@ class UnixDomainSocket_test : public Test
 const size_t UnixDomainSocket_test::MaxMsgSize = UnixDomainSocket::MAX_MESSAGE_SIZE;
 constexpr uint64_t UnixDomainSocket_test::MaxMsgNumber;
 
-TEST_F(UnixDomainSocket_test, NonBlockingModeNotSupported)
+TEST_F(UnixDomainSocket_test, UnlinkEmptySocketNameLeadsToInvalidChannelNameError)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:128]: Aren't this tests for all IPC channel implementations like the message queue, in order to ensure similar behavior?

File: iceoryx_examples/iceperf/uds.cpp [hunk 4]
```diff
@@ -90,7 +90,7 @@ void UDS::initFollower() noexcept
 void UDS::init() noexcept
 {
     // init subscriber
-    iox::posix::posixCall(socket)(AF_LOCAL, SOCK_DGRAM, 0)
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_examples/iceperf/uds.cpp:93]: If we do not use this abstraction then this example will not run in windows. In general I would agree but here it is just a parameter forwarding - a teeny tiny overhead. Could you live with it or do you think that this would effect the outcome of the benchmark in a measurable way?

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp [hunk 5]
```diff
@@ -133,6 +128,16 @@ UnixDomainSocket& UnixDomainSocket::operator=(UnixDomainSocket&& other) noexcept
...
+    {
+        return cxx::error<IpcChannelError>(IpcChannelError::INVALID_CHANNEL_NAME);
+    }
+
+    if (UdsName_t().capacity() < name.size() + UdsName_t(PATH_PREFIX).size())
+    {
+        return cxx::error<IpcChannelError>(IpcChannelError::INVALID_CHANNEL_NAME);
+    }
+
     return unlinkIfExists(NoPathPrefix, UdsName_t(PATH_PREFIX).append(iox::cxx::TruncateToCapacity, name));
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:141]: This is correct and it goes for the user. The `PATH_PREFIX` is set in the header and per default always applied but if the user would like to create an uds in its current folder they can user `unlinkIfExists(NoPathPrefix, "mySocketFile");`

This here is just the implementation avoids code duplication.

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 1 above)
Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:212]: You are right, but I also have to reduce the `MAX_MESSAGE_SIZE` by the terminator size since it is always added.

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 2 above)
Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: There is only the requirement that you provide some kind of filename, therefore the name is not allowed to be empty. The `UdsName_t` is an alias for `cxx::string<LONGEST_VALID_NAME>` therefore the greater than check is done via the type.

A socket file name with one letter is allowed. Therefore we can reduce the check to `name.empty()`.

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp (hunk 3 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:128]: The unlink checks are checks for the ipc channel but were not really done there. Therefore, I created them here specific for the unix domain socket where I created sockets etc.

The unix domain socket also offers features which are not used by the IPC channel and I wanted to clearly test them in a unit test.

File: iceoryx_examples/iceperf/uds.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_examples/iceperf/uds.cpp:93]: Well, I don't mind. You have to ask budrus.

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 6]
```diff
@@ -69,45 +75,282 @@ class UnixDomainSocket_test : public Test
...
+    for (uint64_t i = 0U; i < UnixDomainSocket::LONGEST_VALID_NAME - strlen(UnixDomainSocket::PATH_PREFIX) + 1; ++i)
+    {
+        longSocketName.append(cxx::TruncateToCapacity, "o");
+    }
+    auto ret = UnixDomainSocket::unlinkIfExists(longSocketName);
+    ASSERT_TRUE(ret.has_error());
+    EXPECT_THAT(ret.get_error(), Eq(IpcChannelError::INVALID_CHANNEL_NAME));
+}
+
+TEST_F(UnixDomainSocket_test, UnlinkExistingSocketLeadsIsSuccessful)
```
Comment (FerdinandSpitzschnueffler) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:154]: rename to UnlinkExistingSocketIsSuccessful

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: So `"/"` is a valid name ;)

Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: Of course ... ;) I think in Windows it actually is maybe?!.
But I can extend the test that only the following characters are allowed `[a-zA-Z0-9.,-]` but it wasn't there in the first place...

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 7]
```diff
@@ -69,45 +75,282 @@ class UnixDomainSocket_test : public Test
...
+    EXPECT_THAT(ret.get_error(), Eq(IpcChannelError::INVALID_CHANNEL_NAME));
+}
+
+TEST_F(UnixDomainSocket_test, UnlinkTooLongSocketNameWithPathPrefixLeadsToInvalidChannelNameError)
+{
+    UnixDomainSocket::UdsName_t longSocketName;
+    for (uint64_t i = 0U; i < UnixDomainSocket::LONGEST_VALID_NAME - strlen(UnixDomainSocket::PATH_PREFIX) + 1; ++i)
+    {
+        longSocketName.append(cxx::TruncateToCapacity, "o");
+    }
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:148]: maybe we need a `string::insert` method

**Review by FerdinandSpitzschnueffler — DISMISSED** [review]

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 8]
```diff
@@ -69,45 +75,282 @@ class UnixDomainSocket_test : public Test
...
 
-TEST_F(UnixDomainSocket_test, SendingOnServerLeadsToError)
+void sendOnServerLeadsToError(const sendCall_t& send)
 {
     cxx::string<10> message{"Foo"};
-    Duration duration{1_ms};
-    auto result = server.timedSend(message, duration);
+    auto result = send(message);
     EXPECT_TRUE(result.has_error());
     ASSERT_THAT(result.get_error(), Eq(IpcChannelError::INTERNAL_LOGIC_ERROR));
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:178]: This does not work as intended. The `ASSERT_THAT` returns from this function but on the caller side the execution is continued.
This will terminate the test process.
```cpp
void foo() {
    ASSERT_TRUE(false);
}

TEST(Foo, foo) {
    foo();
    auto n = 1/0;
}
```
Currently this function is not used in a harmful way, though

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 9]
```diff
@@ -69,45 +75,282 @@ class UnixDomainSocket_test : public Test
...
 }
 
-TEST_F(UnixDomainSocket_test, ReceivingOnClientLeadsToError)
+TEST_F(UnixDomainSocket_test, TimedSendOnServerLeadsToError)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:181]: Can you add the reason why it leads to an error?

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 10]
```diff
@@ -69,45 +75,282 @@ class UnixDomainSocket_test : public Test
...
+    sendOnServerLeadsToError([&](auto& msg) { return server.send(msg); });
+}
+
+void successfulSendAndReceive(const std::string& message, const sendCall_t& send, const receiveCall_t& receive)
+{
+    ASSERT_FALSE(send(message).has_error());
+
+    auto receivedMessage = receive();
+    ASSERT_FALSE(receivedMessage.has_error());
+    EXPECT_EQ(message, *receivedMessage);
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:197]: Theoretically we do not yet know if our abstraction is using the sockets. It's possible to use a global variable for data transfer. Can we have one test using the POSIX API for send and receive directly?

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 11]
```diff
@@ -69,45 +75,282 @@ class UnixDomainSocket_test : public Test
...
+
+void receivingOnClientLeadsToError(const receiveCall_t& receive)
+{
+    auto result = receive();
     EXPECT_TRUE(result.has_error());
     ASSERT_THAT(result.get_error(), Eq(IpcChannelError::INTERNAL_LOGIC_ERROR));
 }
-} // namespace
+
+TEST_F(UnixDomainSocket_test, ReceivingOnClientLeadsToErrorWithReceive)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:313]: please add the cause of the error

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: You touched it ;)

That check is not sufficient since you can have `/tmp/mysocket`. The previous check was also not sufficient since this should also not be allowed `/////`. Maybe we just need to check if the last character isn't a path separator?

Comment (elBoberido) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: ... oh, and we need tests for this

File: iceoryx_examples/iceperf/uds.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_examples/iceperf/uds.cpp:93]: @budrus said its alright!

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp [hunk 12]
```diff
@@ -274,7 +278,18 @@ UnixDomainSocket::timedReceive(const units::Duration& timeout) const noexcept
...
+#if defined(__APPLE__)
+    if (tv.tv_sec != 0 || tv.tv_usec != 0)
+    {
+        std::cerr
+            << "socket: \"" << m_name
+            << "\", timedReceive with a timeout != 0 is not supported on MacOS. timedReceive will behave like receive "
+               "instead."
+            << std::endl;
+    }
+#endif
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:290]: You are completely right, but this would be a more complex task since the timeout is not directly provided to the function, it is set in a call before via `setsockopt`.
I created an issue for this: https://github.com/eclipse-iceoryx/iceoryx/issues/823

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp (hunk 7 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:148]: @FerdinandSpitzschnueffler yes we do. Also it would be nice if we could `concatenate`, `append` simple char's.

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp (hunk 8 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:178]: You are right but here an `EXPECT_THAT` suffices.

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp (hunk 10 above)
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:197]: You are right, this would make send ... (for all posix wrapper :( I see the dawn of a new issue)

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp (hunk 7 above)
Comment (FerdinandSpitzschnueffler) [reply iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:151]: Added these methods to https://github.com/eclipse-iceoryx/iceoryx/issues/208
Feel free to add more.

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:290]: can this warning be printed in the `iox_setsockopt` implementation in the mac platform?

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp (hunk 10 above)
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:202]: It has to be done once, though. When we know that e.g. send works, we can use that to test timedReceive, etc.

**Review by elBoberido — COMMENTED** [review]
Can you please improve the naming on the two tests I mentioned in my last PR? Except that, this looks good

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: It seems this will be fixed with the next PR

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:290]: That was my idea. The warning stays but is then printed via platforms.

File: iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp (hunk 2 above)
Comment (elfenpiff) [reply iceoryx_hoofs/source/posix_wrapper/unix_domain_socket.cpp:526]: Is already fixed with tests n stuff in the following PR: https://github.com/eclipse-iceoryx/iceoryx/pull/820

File: iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp [hunk 13]
```diff
@@ -69,45 +75,289 @@ class UnixDomainSocket_test : public Test
...
     cxx::string<10> message{"Foo"};
-    Duration duration{1_ms};
-    auto result = server.timedSend(message, duration);
+    auto result = send(message);
     EXPECT_TRUE(result.has_error());
-    ASSERT_THAT(result.get_error(), Eq(IpcChannelError::INTERNAL_LOGIC_ERROR));
+    EXPECT_THAT(result.get_error(), Eq(IpcChannelError::INTERNAL_LOGIC_ERROR));
+}
+
+TEST_F(UnixDomainSocket_test, TimedSendOnServerLeadsToError)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_unix_domain_sockets.cpp:186]: Oh, how about naming this `ServerCanOnlyReceiveThereforeTimedSendLeadsToError` and similar with the other 3 tests?

**Review by elBoberido — APPROVED** [review]
If you want to keep the names, you can merge this

**Review by FerdinandSpitzschnueffler — APPROVED** [review]
